### PR TITLE
Use wide-char version of CreateFile API

### PR DESCRIPTION
**Pull request by Jinming-Hu**
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2302
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2303

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Review by antkmsft — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
@ahsonkhan You found this issue, I'd like to get your approval before merging this PR and close those two issues.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (CaseyCarter) [inline sdk/storage/azure-storage-common/src/file_io.cpp:34]: I recommend passing `filename.size()` instead of `-1` so the caller gets embedded `\0` characters if they ask for them instead of silently using the prefix of `filename` that precedes any such `\0` character.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (CaseyCarter) [inline sdk/storage/azure-storage-common/src/file_io.cpp:39]: `sizeNeeded` includes the terminating `L'\0'` when the `cbMultiByte` argument to `MultiByteToWideChar` is `-1` as it is currently. Consequently, this string will be one character longer than necessary and have two `L'\0'` characters at the end. If you don't apply my suggestion above, I'd change this to `filenameW(sizeNeeded - 1, L'\0')`.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (CaseyCarter) [inline sdk/storage/azure-storage-common/src/file_io.cpp:95]: Ditto my above comments. (I'd extract this common sequence into a function that widens strings for filenames.)

**Review by CaseyCarter — APPROVED** [review]
I'm approving _and_ suggesting changes since I believe this is correct but could be improved.

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:39]: Hi @CaseyCarter , I tested with code

```C++
  std::string str = "abc";
  int sizeNeeded = MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS, str.data(), -1, nullptr, 0);
```

if it's `-1`, the return value is `4`. If it's `str.length()`, the return value is `3`.

So if I apply your suggestion above, this line should be `std::wstring filenameW(sizeNeeded + 1, L'\0')`. Is that correct?

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp]: I appended a commit to fix this.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (CaseyCarter) [inline sdk/storage/azure-storage-common/src/file_io.cpp:34]: Should we check the implied precondition that `filename.length() <= INT_MAX`? (Also on line 83.)

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (CaseyCarter) [inline sdk/storage/azure-storage-common/src/file_io.cpp:39]: The `+ 1` is unnecessary here. `std::wstring` (as for all specializations of `std::basic_string`) is always null-terminated, and the null terminator isn't included in the size of the string. The effect of the `+ 1` here is that there are two `L'\0'` characters at the end of this string, one that is included in `filename.length()` and one that is not. (Also on line 88.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:34]: I don't think filename longer than `INT_MAX` makes any sense on any OS or filesystem. So I don't think it's necessary

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp:39]: Yeah, I agree. But when we pass the string to another function, the function reads the characters in the string one by one, until meets a `\0`. The function needs to actually **read** the `\0` to know it is a terminator.

Isn't this out-of-range access an undefined behavior?

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (BillyONeal) [reply sdk/storage/azure-storage-common/src/file_io.cpp:34]: The function doesn't have to succeed if that happens; you can check and then either abort or throw.

File: sdk/storage/azure-storage-common/src/file_io.cpp
Comment (BillyONeal) [reply sdk/storage/azure-storage-common/src/file_io.cpp]: No, because basic_string already puts a null there. http://eel.is/c++draft/basic.string#general-3

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/src/file_io.cpp]: fixed

**Pull request merged**
